gcs/constraints/smart_table.cc

```diff
@@ -0,0 +1,642 @@
+                },
+                [&](const UnarySetEntry &) {
+                    entry_tree[current_level].emplace_back(edge);
+                },
+                [&](const UnaryValueEntry &) {
+                    entry_tree[current_level].emplace_back(edge);
+                }}
+                .visit(edge);
+        }
+    }
+
+    auto build_forests(SmartTuples & tuples) -> vector<SmartTable::Forest>
+    {
+        vector<SmartTable::Forest> forests{};
+        for (const auto & current_tuple : tuples) {
+            unordered_map<IntegerVariableID, bool> node_visited;
+            unordered_map<IntegerVariableID, vector<SmartEntry>> adjacent_edges;
+
+            // Get all the vars in the tuple and record adjacencies
+            for (const auto & entry : current_tuple) {
+                overloaded{
+                    [&](const BinaryEntry & binary_entry) {
+                        node_visited[binary_entry.var_1] = false;
+                        node_visited[binary_entry.var_2] = false;
+                        adjacent_edges.try_emplace(binary_entry.var_1, vector<SmartEntry>{});
+                        adjacent_edges.try_emplace(binary_entry.var_2, vector<SmartEntry>{});
+                        adjacent_edges[binary_entry.var_1].emplace_back(binary_entry);
+                        adjacent_edges[binary_entry.var_2].emplace_back(binary_entry);
+                    },
+                    [&](const UnaryValueEntry & unary_val_entry) {
+                        node_visited[unary_val_entry.var] = false;
+                        adjacent_edges.try_emplace(unary_val_entry.var, vector<SmartEntry>{});
+                        adjacent_edges[unary_val_entry.var].emplace_back(unary_val_entry);
+                    },
+                    [&](const UnarySetEntry & unary_set_entry) {
+                        node_visited[unary_set_entry.var] = false;
+                        adjacent_edges.try_emplace(unary_set_entry.var, vector<SmartEntry>{});
+                        adjacent_edges[unary_set_entry.var].emplace_back(unary_set_entry);
+                    }}
+                    .visit(entry);
+            }
+
+            SmartTable::Forest forest{};
+            for (auto & var_pair : node_visited) {
+                auto & var = var_pair.first;
+                auto & visited = var_pair.second;
+                if (visited) continue;
+                vector<vector<SmartEntry>> entry_tree;
+                entry_tree.emplace_back();
+                // Recursively build the tree starting from this node
+                build_tree(var, 0, entry_tree, node_visited, adjacent_edges);
+                forest.emplace_back(entry_tree);
+            }
+
+            forests.emplace_back(forest);
+        }
+        return forests;
+    }
+}
+
+// For PB model
+auto make_binary_entry_flag(State & state, Propagators & propagators, const IntegerVariableID & var_1, const IntegerVariableID & var_2, const ConstraintType & c) -> ProofFlag
```

Comment:
Should this be in the anonymous namespace?

Reply:
```suggestion
[[nodiscard]] auto make_binary_entry_flag(State & state, Propagators & propagators, const IntegerVariableID & var_1, const IntegerVariableID & var_2, const ConstraintType & c) -> ProofFlag
```

Reply:
Fixed in c62e8914eb9e01df2a147f01713ff505d7e905e0